pyogrio/_io.pyx

```diff
@@ -1678,6 +1678,117 @@ cdef infer_field_types(list dtypes):
 
     return field_types
 
+def ogr_write_arrow(
+    obj,
+    str path
+):
+    IF CTE_GDAL_VERSION < (3, 8, 0):
+        raise RuntimeError("Need GDAL>=3.8 for Arrow write support")
+
+    stream_capsule = obj.__arrow_c_stream__()
+    return write_arrow_stream_capsule(stream_capsule)
+
+cdef OGRErr write_arrow_stream_capsule(OGRLayerH destLayer, object capsule):
+    cdef ArrowArrayStream stream
+    cdef ArrowSchema schema
+    cdef ArrowArray array
+
+    stream = PyCapsule_GetPointer(capsule, "arrow_array_stream")
+    if not stream:
+        raise RuntimeError("Accessing PyCapsule pointer named 'arrow_array_stream' failed.")
```

Comment:
It's probably fine to check for a NULL stream, but that the message can say something about "no valid stream" (the accessing of the pointer itself will have succeeded at this point, because if that fails, `PyCapsule_GetPointer` already raises an error which cython translates to a Python exception)